fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/planning/BooleanPredicateNormalizer.java

```diff
@@ -0,0 +1,408 @@
+            if (failIfTooLarge) {
+                throw new NormalFormTooLargeException(Objects.requireNonNull(predicate));
+            } else {
+                return Optional.empty();
+            }
+        } else {
+            final List<Collection<? extends QueryPredicate>> majorOfMinor = toNormalized(Objects.requireNonNull(predicate), false);
+            trimTerms(majorOfMinor);
+            return Optional.of(mode.majorWithChildren(majorOfMinor.stream().map(mode::minorWithChildren).collect(Collectors.toList())));
+        }
+    }
+
+    private boolean needsNormalize(@Nullable final QueryPredicate predicate) {
+        if (predicate == null) {
+            return false;
+        }
+
+        return isBooleanPredicate(predicate) &&
+               StreamSupport.stream(predicate.getChildren().spliterator(), false).anyMatch(this::isBooleanPredicate);
+    }
+
+    private boolean shouldNormalize(@Nullable final QueryPredicate predicate) {
+        try {
+            return getNormalizedSize(predicate) <= sizeLimit;
+        } catch (ArithmeticException e) {
+            // Our computation caused an integer overflow so the DNF is _definitely_ too big.
+            return false;
+        }
+    }
+
+    private boolean isBooleanPredicate(@Nullable final QueryPredicate predicate) {
+        return predicate instanceof AndOrPredicate ||
+               predicate instanceof NotPredicate;
+    }
+
+    int getNormalizedSize(@Nullable final QueryPredicate predicate) {
+        if (predicate == null) {
+            return 0;
+        }
+        return toNormalizedSize(predicate, false);
+    }
+
+    private int toNormalizedSize(@Nonnull final QueryPredicate predicate, final boolean negate) {
+        if (mode.instanceOfMinorClass(predicate)) {
+            final List<? extends QueryPredicate> children = ((AndOrPredicate)predicate).getChildren();
+            return negate ? majorToNormalizedSize(children, true) : minorToNormalizedSize(children, false);
+        } else if (mode.instanceOfMajorClass(predicate)) {
+            final List<? extends QueryPredicate> children = ((AndOrPredicate)predicate).getChildren();
+            return negate ? minorToNormalizedSize(children, true) : majorToNormalizedSize(children, false);
+        } else if (predicate instanceof NotPredicate) {
+            return toNormalizedSize(((NotPredicate)predicate).getChild(), !negate);
+        } else {
+            return 1;
+        }
+    }
+
+    private int majorToNormalizedSize(@Nonnull final List<? extends QueryPredicate> children, final boolean negate) {
+        return children.stream().mapToInt(p -> toNormalizedSize(p, negate)).reduce(0, Math::addExact);
+    }
+
+    private int minorToNormalizedSize(@Nonnull final List<? extends QueryPredicate> children, final boolean negate) {
+        return children.stream().mapToInt(child -> toNormalizedSize(child, negate)).reduce(1, Math::multiplyExact);
+    }
+
+    /**
```

Comment:
The Javadoc on the next few methods and the comment in `trimTerms` assume DNF mode. I don't know how to rewrite them in a way that's still comprehensible. Also, in the trim case, the implication is in the other direction, from subset to superset, and that's what's needed.

Reply:
Will do the java doc changes.

In terms of which way the implication goes, wouldn't it be the same as both

`a v (a ^ b) = a`
and
`a ^ (a v b) = a`

or am I misunderstanding your comment?

Reply:
Oh -- do you mean the way the comment in the method is written?

Reply:
Yes, exactly. `(a ∧ b ∧ c) ⇒ (a ∧ b)` and `(a ∨ b) ⇒ (a ∨ b ∨ c)`.